### DISCUSSION TITLE
[Firestore] admin.firestore.Timestamp.now() fails with firebase-admin 11, works in 10.3 (emulator)

### DISCUSSION THREAD

**Discussion by tofsjonas** [firebase/firebase-admin-node#1959] [General]
### [REQUIRED] Step 2: Describe your environment

  * Operating System version: macOS Monterey 12.4
  * Firebase SDK version: 11.3.0
  * Firebase Product: Firestore
  * Node.js version: v16.14.2
  * NPM version: 8.15.0

tsconfig.json

``` json
{
  "compilerOptions": {
    "lib": ["ES2020"],
    "module": "commonjs",
    "target": "ES2020",
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "outDir": "dist",
    "sourceMap": true,
    "strict": true,
    "allowSyntheticDefaultImports": true
  },
  "compileOnSave": true,
  "include": ["src"],
  "exclude": ["node_modules", "**/node_modules/*"]
}

```

package.json

```json
{
  "name": "functions",
  "scripts": {
    "build": "rm -rf dist && tsc",
    "watch": "rm -rf dist && npx tsc -w --skipLibCheck --noEmitOnError",
    "serve": "firebase emulators:start --import=./.cache --export-on-exit",
    "deploy": "firebase deploy --only hosting",
    "logs": "firebase functions:log"
  },
  "engines": {
    "node": "16"
  },
  "main": "dist/index.js",
  "dependencies": {
    "@firebase/app-compat": "^0.1.29",
    "axios": "^0.27.2",
    "cors": "^2.8.5",
    "firebase-admin": "^11.0.0",
    "firebase-functions": "^3.22.0",
    "stripe": "^9.14.0"
  },
  "devDependencies": {
    "@firebase/app-types": "^0.7.0",
    "@types/cors": "^2.8.12",
    "@types/stripe-v3": "^3.1.27",
    "firebase-functions-test": "^2.2.0",
    "jest": "^28.1.3",
    "typescript": "^4.7.4"
  },
  "private": true
}

```
### [REQUIRED] Step 3: Describe the problem

`admin.firestore.Timestamp.now()` throws **TypeError: Cannot read properties of undefined (reading 'now')** (~same with `admin.firestore.increment(1)`). 

In emulator, I haven't dared try it "live".

#### Steps to reproduce / Relevant Code:

test.ts

``` javascript
import * as functions from 'firebase-functions'
import * as admin from 'firebase-admin'
export const test = functions.region('europe-west1').https.onRequest(async (request, response) => {
  const res = await admin.firestore().collection('mytable').add({
    createdAt: admin.firestore.Timestamp.now(),
  })
  response.json({
    id: res.id,
  })
})
```

Result:

`TypeError: Cannot read properties of undefined (reading 'now')`

[a comment by an automated account is not shown]

**Comment by Joebayld** [comment]
It appears I'm having the same issue with the `delete` FieldValue. I get this error `TypeError: Cannot read properties of undefined (reading 'delete')`. 

This looks to be yet another bug that has happened with firebase code releases. It seems like something always breaks on major releases these days.

**Comment by AkshayChhajed1998** [comment]
I am having same issues on emulator with  `new admin.firestore.Timestamp(s,ns)` says Timestamp() is not the constructor  , but works good live

**Comment by milo-** [comment]
Also experiencing the same issue, and `firestore.v1.FirestoreAdminClient` is also undefined

**Comment by magician11** [comment]
I rolled back to `v10.3.0` and `admin.firestore.Timestamp.now()` works again.

**Comment by lahirumaramba** [comment]
Thank you for reporting this issue. This might have been caused by the Firestore version bump in Admin SDK `v11`.
In the meantime could try adding `import { Timestamp } from 'firebase-admin/firestore'` as a workaround? It seemed to have worked in #1793

**Comment by tofsjonas** [comment]
When I do  `import { Timestamp } from 'firebase-admin/firestore'` I get this:

> Unable to resolve path to module 'firebase-admin/firestore'.eslint [import/no-unresolved](https://github.com/import-js/eslint-plugin-import/blob/v2.26.0/docs/rules/no-unresolved.md)

😢

This seems to work: `import type { DocumentReference, DocumentData, Timestamp } from 'firebase-admin/firestore'` 🤔🤯

**Comment by Joebayld** [comment]
> Thank you for reporting this issue. This might have been caused by the Firestore version bump in Admin SDK `v11`.
> In the meantime could try adding `import { Timestamp } from 'firebase-admin/firestore'` as a workaround? It seemed to have worked in #1793

What about for the delete field value? 

Any idea when this can be patched?

**Comment by lahirumaramba** [comment]
It seems like as we [move on to the modular SDK](https://firebase.google.com/docs/admin/migrate-node-v10#node.js) we should import the types as needed.
The following seems to work for me:

**Node.js**
```
const { initializeApp } = require('firebase-admin/app');
const { getFirestore, Timestamp, FieldValue } = require('firebase-admin/firestore');

initializeApp();

const f = getFirestore();
const t = Timestamp.now();
const fv = FieldValue.delete();
```

**TypeScript**
```
import { getFirestore, Timestamp, FieldValue, Firestore } from 'firebase-admin/firestore';
import { initializeApp } from 'firebase-admin/app';

initializeApp()

const f: Firestore = getFirestore();
const t: Timestamp = Timestamp.now();
const fv: FieldValue = FieldValue.delete();
```

**TypeScript** (legacy - using the global `admin` namespace is deprecated) 
```
import * as admin from 'firebase-admin';

admin.initializeApp();

const f = admin.firestore();
const t = admin.firestore.Timestamp.now();
const fv = admin.firestore.FieldValue.delete();
```

**tsconfig.json**
```
{
  "compilerOptions": {
    "module": "commonjs",
    "esModuleInterop": true,
    "target": "es6",
    "moduleResolution": "node",
    "sourceMap": true,
    "outDir": "dist"
  },
  "lib": ["es2015"]
}
```

**package.json**
```
{
  "name": "repro",
  "version": "1.0.0",
  "main": "dist/test.js",
  "scripts": {
    "start": "tsc && node dist/test.js"
  },
  "engines": {
    "node": "16"
  },
  "devDependencies": {
    "typescript": "^4.7.4"
  },
  "dependencies": {
    "firebase-admin": "^11.0.1"
  }
}
```

```
├─┬ firebase-admin@11.0.1
│ ├── @fastify/busboy@1.1.0
│ ├── @firebase/database-compat@0.2.4
│ ├── @firebase/database-types@0.9.12
│ ├── @google-cloud/firestore@5.0.2
│ ├── @google-cloud/storage@6.3.0
│ ├── @types/node@18.6.5
│ ├── jsonwebtoken@8.5.1
│ ├── jwks-rsa@2.1.4
│ ├── node-forge@1.3.1
│ └── uuid@8.3.2
└── typescript@4.7.4
```

**Reply by waldothedeveloper** [reply]
gold comment here! I was trying 
```javascript
admin.firestore.Timestamp.now();
``` 

in a v.11 and that wasn't working...thank you!

**Reply by x5engine** [reply]
should be added to documentation tbh

**Reply by seth-stitik** [reply]
why could i not find this in the docs?? thx for sharing. work for me w typescript

**Comment by tofsjonas** [comment]
Ah.

The problem was (is) `eslint-plugin-import` 😢

This will throw lint errors: 

```javascript
import { getFirestore, Timestamp, FieldValue, Firestore } from 'firebase-admin/firestore';
import { initializeApp } from 'firebase-admin/app';
```
> Unable to resolve path to module 'firebase-admin/app'. eslint [import/no-unresolved](https://github.com/import-js/eslint-plugin-import/blob/v2.26.0/docs/rules/no-unresolved.md)

whereas this will not (which I only tried because I copy/pasted your code):

```javascript
const { initializeApp } = require('firebase-admin/app');
const { getFirestore, Timestamp, FieldValue } = require('firebase-admin/firestore');
```

The "solution" is to not use the plugin anymore. But I would like to..

Do you have any idea what that's about?

Failing eslintrc.js:
```javascript
  extends: [
    'eslint:recommended',
    'plugin:import/recommended',
    'plugin:import/typescript',
```
Passing eslintrc.js:
```javascript
  extends: [
    'eslint:recommended',
    // 'plugin:import/recommended',
    // 'plugin:import/typescript',
```

**Reply by lahirumaramba** [reply]
`eslint-plugin-import` does not currently support `exports` in `package.json`. Please see my reply below https://github.com/firebase/firebase-admin-node/discussions/1959#discussioncomment-4049965

**Comment by Joebayld** [comment]
> It seems like as we [move on to the modular SDK](https://firebase.google.com/docs/admin/migrate-node-v10#node.js) we should import the types as needed. The following seems to work for me:
> 

So is this the way of using the new version of the library with typescript? I can update my code accordingly, just want to make sure this is the correct solution or just a temporary fix. 

Also, can the library be fixed so the old way isn't an option in typescript anymore? Right now the error is thrown only at runtime but the types think it's okay as is (which it obviously isn't).

**Comment by lahirumaramba** [comment]
@Joebayld yes, **if you are using the modular SDK** then this is the correct way to import the types. For the issue you are facing with the previous syntax, could you provide us with a minimal repro by any chance? Thank you!

**Comment by Joebayld** [comment]
> @Joebayld yes, **if you are using the modular SDK** then this is the correct way to import the types. For the issue you are facing with the previous syntax, could you provide us with a minimal repro by any chance? Thank you!

I think I see the issue then. I was using the older style imports (which worked in v10 and v11, but had the runtime issue).

```js 
import * as admin from "firebase-admin";

admin.firestore.FieldValue.delete()
```

With the above code on the latest version, this appears to be fine in typescript but doesn't work. I switched everything to the new modular imports and it seems to be okay.

**Comment by JFGHT** [comment]
With the non-modular approach, we are having the same problem.

**Comment by ansells** [comment]
> Ah.
> 
> The problem was (is) `eslint-plugin-import` 😢
> 
> This will throw lint errors:
> 
> ```js
> import { getFirestore, Timestamp, FieldValue, Firestore } from 'firebase-admin/firestore';
> import { initializeApp } from 'firebase-admin/app';
> ```
> 
> > Unable to resolve path to module 'firebase-admin/app'. eslint [import/no-unresolved](https://github.com/import-js/eslint-plugin-import/blob/v2.26.0/docs/rules/no-unresolved.md)
> 

I'm having this same issue and had to explicitly ignore the rule for this import:

```js
// eslint-disable-next-line import/no-unresolved 
import {Timestamp} from "firebase-admin/firestore";
```

**Comment by pascalbe-dev** [comment]
I also stumbled across that issue. Using the modular imports fixed the issue. Thanks for the hint :) 

I think - since this is a breaking change in usage with the emulator - this should be noted in the release notes.

**Reply by tannerstern** [reply]
> I also stumbled across that issue. Using the modular imports fixed the issue. Thanks for the hint :)
> 
> I think - since this is a breaking change in usage with the emulator - this should be noted in the release notes.

We have recently run into this problem. Our live code works fine, but any namespace usages like `admin.database.ServerValue.TIMESTAMP` or `admin.firestore.Timestamp.now()` throw the `TypeError` in the emulator.

I understand (and agree with conceptually) the direction that Firebase is heading towards modular access, but if namespaced access is still supported, I would think these should still work across the Firebase ecosystem, including in the emulator?

Is that faulty thinking? I suppose we could downgrade to v10.3, but that doesn't seem right to me.

**Reply by johnnyoshika** [reply]
I'm stuck on an older version because of this. I'd like to upgrade, but it's a big undertaking to convert our codebase.

**Reply by lahirumaramba** [reply]
Thanks for the feedback, folks. I am sorry that you are facing this issue and I understand your frustration!

To help us investigate this better, could you confirm a few things for me, please?

1. Does this occur when you use `require` or `import` the global namespace as `import * as admin from 'firebase-admin'`?
2. Are you able to provide us with a minimal repro?
Thanks!

**Reply by tannerstern** [reply]
Hey thanks @lahirumaramba, appreciate you looking into this.

1. This occurs when we `require` the namespace: `const admin = require('firebase-admin');`
2. I could spin something up if you like, but I posted [an issue in the firebase-tools repo](https://github.com/firebase/firebase-tools/issues/5460) (thinking it may be related to the emulator?) and there described what I think are the minimal steps to reproduce the issue, with the required versions of `firebase-functions`, `firebase-admin`, and `firebase-tools`. If you'd like me to set up a repo instead, just let me know.

Thanks again!

**Reply by lahirumaramba** [reply]
I was testing a different issue (https://github.com/firebase/firebase-admin-node/issues/1488#issuecomment-1462935632) and I think it may be related. I was able to reproduce the issue you are seeing on the emulator and explicitly using the `default` export seem to fix it for me.

```
const admin = require('firebase-admin').default;

admin.firestore.Timestamp.now();
```

Unfortunately, I am unsure why this happens so I will have to investigate further to find a proper fix. In the meantime, could you use the `.default` syntax in the emulator mode? Thanks, folks!

**Comment by pascalbe-dev** [comment]
Running my jest tests for code, which imports `getFirestore` the modular way now fails. 
It is unhappy with the following statement in the `/node_modules/firebase-admin/lib/esm/firestore/index.js` file. 

> import mod from "../../firestore/index.js";

It says: `SyntaxError: Cannot use import statement outside a module`

Has anyone experienced this issue as well or knows how to fix it?

**Comment by mariamghalleb** [comment]
> Also experiencing the same issue, and `firestore.v1.FirestoreAdminClient` is also undefined

`import * as firestore from 'firebase-admin/firestore'
`
fixed it for me

**Comment by abMarlim** [comment]
@ lahirumaramba you are a lifesave, thanks <3

**Comment by lahirumaramba** [comment]
Since it seems like switching to the modular SDK is the proper fix here I will convert this issue to a discussion. That will hopefully help with future issues like #1954

**Comment by readyiman** [comment]
@lahirumaramba did ignore eslint also the solution? because eslint user has this issue
> I'm having this same issue and had to explicitly ignore the rule for this import:
> 
> ```js
> // eslint-disable-next-line import/no-unresolved 
> import {Timestamp} from "firebase-admin/firestore";
> ```

**Reply by lahirumaramba** [reply]
We use `exports` in `package.json` for defining entry points for both `cjs` and `esm` and `eslint-plugin-import` currently does not support [`exports`](https://nodejs.org/api/packages.html#exports). You will have to disable `import/no-unresolved` until there is a fix in `eslint-plugin-import`.

Please see:
[1] https://stackoverflow.com/a/69959606/1526201
[2] https://github.com/import-js/eslint-plugin-import/issues/1810

**Reply by readyiman** [reply]
That [2] looks stalled. 
I ended up removing eslint import plugin, I think it may not needed, no error so far.

**Comment by johnnyoshika** [comment]
We encountered this problem when importing like this using TypeScript:

```
import * as admin from 'firebase-admin';
```

It's been a while now, but if I recall correctly, the problem didn't appear in the Emulator but it did once I deployed to the cloud.

**Reply by Cvmcosta** [reply]
Hey @johnnyoshika, nice running into you here. Just ran into this issue in the emulator. I panicked and went to check production and everything is working fine there. Such weird behavior, i'll migrate to the new recommendation of modular imports.

**Reply by johnnyoshika** [reply]
@Cvmcosta Hello hello! Glad to hear that everything is OK. I'll need to migrate our my company's codebase to modular imports as well.

**Reply by lahirumaramba** [reply]
I ran a few tests locally because of (https://github.com/firebase/firebase-admin-node/issues/1488#issuecomment-1462935632) and it seems `import * as admin from 'firebase-admin'` imports all the exports from the module... so you have to specifically access the `default` export... I am unsure why this happens though.

**Comment by cybex-dev** [comment]
Emulator vs Live deployment for firestore-admin & Timestamp imports with temporary solution.

**Environment:**

> /usr/local/lib
> └── firebase-tools@11.24.0

> /functions
> └── firebase-tools@11.16.1

**OS:** macOs
**NPM:** 9.6.0
**Node:** v14.19.0

---

**Details:**

Using `Timestamp.now()` or `Timetamp.fromDate()` (haven't tested any other functions), emulator and "live" aren't compatible. Regression from `firebase-tools@10.9.2` (possibly earlier)

Using the following on **live** builds, deploys, and executes as expected:

```ts
import {firestore} from "firebase-admin";
import Timestamp = firestore.Timestamp;
```

but emulator(s) do not like it. Emulator function crashes on calling:

```Timestamp.now()```

with stacktrace:

> functions: TypeError: Cannot read property 'now' of undefined
>     at /Users/mac/TestProject/server/functions/lib/controllers/firebase-auth/listeners/listeners.js:35:36
>     at cloudFunction (/Users/mac/TestProject/server/functions/node_modules/firebase-functions/lib/v1/cloud-functions.js:79:23)
>     at /Users/mac/TestProject/server/functions/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:518:16
>     at runFunction (/Users/mac/TestProject/server/functions/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:506:15)
>     at runBackground (/Users/mac/TestProject/server/functions/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:517:11)
>     at processBackground (/Users/mac/TestProject/server/functions/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:501:11)
>     at /Users/mac/TestProject/server/functions/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:686:27
>     at Layer.handle [as handle_request] (/Users/mac/TestProject/server/functions/node_modules/firebase-tools/node_modules/express/lib/router/layer.js:95:5)
>     at next (/Users/mac/TestProject/server/functions/node_modules/firebase-tools/node_modules/express/lib/router/route.js:144:13)
>     at next (/Users/mac/TestProject/server/functions/node_modules/firebase-tools/node_modules/express/lib/router/route.js:140:7)
> ⚠  Your function was killed because it raised an unhandled error.
> 

---

Using the following on **emulator** builds, runs, and executes as expected:

```ts
import {Timestamp} from "firebase-admin/firestore";
```
but does not pass lint checks, see for lint bypass rule: https://github.com/firebase/firebase-admin-node/discussions/1359#discussioncomment-977158

**TL;DR** 

In `${firebase_project}/functions/.eslintrc.js`, add
```js
module.exports = {
    // ...,
    rules: {
        // ...
        'import/no-unresolved':
            [
                'error',
                {
                    ignore: ['^firebase-admin/.+'],
                },
            ]
    }
}
```

**Comment by richard-rowdy-jr** [comment]
@lahirumaramba thank you for your help here, I had a question tied to this since it was a bit hard to find examples of the new modular way, especially with these timestamps and stuff. 
I remember firebase had different examples in git somewhere with the older libraries, do you know if there is a set of code samples now with the modular admins SDK anywhere? just like basics how to do different trigger functions, how to best do imports, use cases, etc
anything would be helpful :pray:

**Comment by Dhinesh-Dhinesh** [comment]
This one works for me

```typescript
import { FieldValue } from 'firebase-admin/firestore';

..........

const data = {
   createdAt: FieldValue.serverTimestamp()
}
```

**Comment by rafael-camara** [comment]
This one works for me

```js
import { initializeApp } from 'firebase/app';
import { getDatabase } from 'firebase/database';
import { Timestamp } from 'firebase/firestore';

const firebaseConfig = {
....
}

const firebaseApp = initializeApp(firebaseConfig);
const database = getDatabase(firebaseApp);

const TIMESTAMP = Timestamp.now().toMillis();

function writeData(data) {
    const databaseRef = ref(database, endpoint);
    const newDataRef = push(databaseRef);

    set(newDataRef, data);
}

const data = {
  .....
  createdAt: TIMESTAMP,
}

....
writeData(data);
```

**Comment by msusman1** [comment]
This one is working for me for the CommonJS environment.

```javascript
const { FieldValue, Timestamp} = require("firebase-admin/firestore");

// ...

const serverTime = FieldValue.serverTimestamp();
await rateLimitRef.set({ timestamp: serverTime });

//...

const now = Timestamp.now();
const cutoff = new Date(now.toDate().getTime() - 24 * 60 * 60 * 1000); // 24 hours ago
const snapshot = await rateLimitRef.where('timestamp', '<=', cutoff).get();
